node/overseer/overseer-gen/proc-macro/src/parse_attr.rs

```diff
@@ -0,0 +1,227 @@
+			})
+		} else {
+			Err(lookahead.error())
+		}
+	}
+}
+
+/// Attribute arguments
+#[derive(Clone, Debug)]
+pub(crate) struct AttrArgs {
+	pub(crate) message_wrapper: Ident,
+	pub(crate) extern_event_ty: Path,
+	pub(crate) extern_signal_ty: Path,
+	pub(crate) extern_error_ty: Path,
+	/// A external subsystem that both consumes and produces messages
+	/// but is not part of the band of subsystems, it's a mere proxy
+	/// to another entity that consumes/produces messages.
+	pub(crate) extern_network_ty: Option<Path>,
+	pub(crate) outgoing_ty: Option<Path>,
+	pub(crate) signal_channel_capacity: usize,
+	pub(crate) message_channel_capacity: usize,
+}
+
+macro_rules! extract_variant {
+	($unique:expr, $variant:ident ; default = $fallback:expr) => {
+		extract_variant!($unique, $variant)
+			.unwrap_or_else(|| { $fallback })
+	};
+	($unique:expr, $variant:ident ; err = $err:expr) => {
+		extract_variant!($unique, $variant)
+			.ok_or_else(|| {
+				Error::new(Span::call_site(), $err)
+			})
+	};
+	($unique:expr, $variant:ident) => {
+		$unique.values()
+			.find_map(|item| {
+				if let OverseerAttrItem:: $variant { value, ..} = item {
+					Some(value.clone())
+				} else {
+					None
+				}
+			})
+	};
+}
+
+impl Parse for AttrArgs {
+	fn parse(input: &ParseBuffer) -> Result<Self> {
+		let items: Punctuated<OverseerAttrItem, Token![,]> = input.parse_terminated(OverseerAttrItem::parse)?;
+
+		let mut unique = HashMap::<std::mem::Discriminant<OverseerAttrItem>, OverseerAttrItem, RandomState>::default();
+		for item in items {
+			if let Some(first) = unique.insert(std::mem::discriminant(&item), item.clone()) {
+				let mut e = Error::new(item.span(), format!("Duplicate definition of overseer generation type found"));
+				e.combine(Error::new(first.span(), "previously defined here."));
+				return Err(e);
+			}
+		}
+
+		let signal_channel_capacity = extract_variant!(unique, SignalChannelCapacity; default = 64_usize);
+		let message_channel_capacity = extract_variant!(unique, MessageChannelCapacity; default = 1024_usize);
+
+		let error = extract_variant!(unique, ExternErrorType; err = "Must declare the overseer error type via `error=..`.")?;
+		let event = extract_variant!(unique, ExternEventType; err = "Must declare the overseer event type via `event=..`.")?;
+		let signal = extract_variant!(unique, ExternOverseerSignalType; err = "Must declare the overseer signal type via `span=..`.")?;
+		let message_wrapper = extract_variant!(unique, MessageWrapperName; err = "Must declare the overseer generated wrapping message type via `gen=..`.")?;
+		let network = extract_variant!(unique, ExternNetworkType);
+		let outgoing = extract_variant!(unique, OutgoingType);
+
+		Ok(AttrArgs {
+			signal_channel_capacity,
+			message_channel_capacity,
+			extern_event_ty: event,
+			extern_signal_ty: signal,
+			extern_error_ty: error,
+			extern_network_ty: network,
+			outgoing_ty: outgoing,
+			message_wrapper,
+		})
+	}
+}
```

Comment:
@bkchr could you take a glance at this if there are further changes mandated? Thanks! Much appreciated.

Reply:
LGTM